test-common/utils.c

```diff
@@ -698,107 +877,136 @@ generate_delta_files (GFile *repo,
   return cmd_result_ensure_ok (&cmd, error);
 }
 
-/**
- * get_last_ref:
- * @ref_to_commit: (element-type utf8 uint): a #GHashTable mapping refs to commit numbers.
- * @wanted_commit_number: the commit number to count down from.
+/* Breadth-first traversal of the commit graph, starting out
+ * at the root note and expanding each node until we get to
+ * the end.
  *
- * Look through the ref_to_commit hashtable (which maps refs to
- * commit numbers) to try and find the last known ref before
- * wanted_commit_number. It handles the case where we have commits
- * N, N - J and don't have (N - 1)..(N - J) in the hashtable.
- *
- * Returns: the last known ref before wanted_commit_number
- */
-static OstreeCollectionRef *
-get_last_ref (GHashTable *ref_to_commit,
-              guint       wanted_commit_number)
+ * The implementation is a little awkward since we need to do
+ * a linear scan on the parents hashtable to figure out the
+ * children of a given commit. */
+gboolean
```

Comment:
This could also be implemented as a forward adjacency list ... I guess there's just more stuff to keep track of then (every node will need to have a pointer array of children). Personally I don't mind the O(V) cost to expand children for each node since this is test code.

Reply:
Indeed, as long as the tests run without timing out, I don’t care about performance. Also, we’re talking about ~10 test commits in the graph.